contract/src/lib.rs

```diff
@@ -404,6 +409,69 @@ impl BtcLightClient {
         }
     }
 
+    fn check_target(&self, block_header: &Header, prev_block_header: &ExtendedHeader) {
+        if prev_block_header.block_height % BLOCKS_PER_ADJUSTMENT != 0 {
+            require!(
+                block_header.bits == prev_block_header.block_header.bits,
+                format!(
+                    "Error: Incorrect bits. Expected bits: {}; Actual bits: {}.",
+                    prev_block_header.block_header.bits, block_header.bits
+                )
+            );
+            return;
+        }
+
+        if prev_block_header.block_height + 1 >= BLOCKS_PER_ADJUSTMENT {
+            if let Some(init_header_hash) = self
+                .mainchain_height_to_header
+                .get(&(prev_block_header.block_height + 1 - BLOCKS_PER_ADJUSTMENT))
+            {
+                let init_extend_header = self
+                    .headers_pool
+                    .get(&init_header_hash)
+                    .unwrap_or_else(|| env::panic_str(ERR_KEY_NOT_EXIST));
+                let actual_time_taken =
+                    prev_block_header.block_header.time - init_extend_header.block_header.time;
+                let last_target = prev_block_header.block_header.target();
+
+                let (mut new_target, new_target_overflow) =
+                    last_target.overflowing_mul(u64::from(actual_time_taken));
+                require!(!new_target_overflow, "new target overflow");
+                new_target = new_target / U256::from(EXPECTED_TIME);
+
+                let (max_target, max_target_overflow) =
+                    last_target.overflowing_mul(MAX_ADJUSTMENT_FACTOR);
+                require!(!max_target_overflow, "max target overflow");
+
+                new_target = min(new_target, max_target)
+                    .max(last_target / U256::from(MAX_ADJUSTMENT_FACTOR));
+
+                let expected_bits = new_target.target_to_bits();
+
+                require!(
+                    expected_bits == block_header.bits,
+                    format!(
+                        "Error: Incorrect target. Expected bits: {:?}, Actual bits: {:?}",
+                        expected_bits, block_header.bits
+                    )
+                );
+                return;
+            }
+        }
+
+        let prev_difficulty = prev_block_header.block_header.work();
+        let current_difficulty = block_header.work();
+
+        require!(
+            prev_difficulty / current_difficulty <= U256::from(MAX_ADJUSTMENT_FACTOR),
+            "Error: The difficulty change exceeds 4 times."
+        );
+        require!(
+            current_difficulty / prev_difficulty <= U256::from(MAX_ADJUSTMENT_FACTOR),
+            "Error: The difficulty change exceeds 4 times."
+        );
```

Comment:
Do we really need this check? 
As I see, the target is calculated and verified in the previous section.

Reply:
now it should be more clear (https://github.com/Near-One/btc-light-client-contract/commit/cb63c1fc39d6584f81a41098a61bdd24b4077479)

We do this checks only in case, when we don't have all the blocks in the beginning

Reply:
fixed: https://github.com/Near-One/btc-light-client-contract/pull/64/commits/c6215e245488db69dcd12516a810e1ae3696b394